dafny-lang/ide-vscode
# Add verification management UI

### Changes

- Using the VSCode test API, provide an overview of available verification tasks, their running status, their outcome, and options to manually start and cancel verification task runs.
- Change the status bar implementation so the messages are also updated by manually triggered verification. Also the messages are slightly altered. When verification succeeds, if not everything was verified, then the number of skipped declarations is shown. When verification fails, the number of declarations that could not be proven is shown.
- Change verification gutter icons so they skip lines that have a test run icon from the VSCode testing API on them.

### Testing

- Manually tested the changes described above.
  - Test backwards compatibility mode
    - When using Dafny 3.7.3, the behavior is the same as it was before this PR.
    - When using the latest nightly and when using a custom LSP server, the new behavior activates .
  - Test that the new behavior works well when using
    - Never verify
    - Verify on change
    - Verify on save
  - Test with and without gutter icons enabled
- Will add automatic tests as well

### Screenshots

The following screenshots shows the effect of having "automatic verification" set to "never", having configured 2 cores for verification, and then manually clicking "run all" in the testing panel. First the two easy proofs m5 and fib verify quickly, then two slow proofs m3 and m4 are verifying, while m1 and m2 are still queued, and this moment is captured by the screenshot.
<img width="1107" alt="image" src="https://user-images.githubusercontent.com/3121201/181475006-6241cf13-17d7-4936-8af1-0b9769baa20d.png">

Once all the verification has completed, we see:
<img width="1109" alt="image" src="https://user-images.githubusercontent.com/3121201/181474650-8278c7dc-d2f4-4f6f-80f0-c98650ddbcdf.png">

When using module and classes, the hierarchy is reflected in the test view:
<img width="1061" alt="image" src="https://user-images.githubusercontent.com/3121201/181775022-93a0cbc4-13e8-4725-8e27-59775c5de6b8.png">

Alternatively, a list view can be shown:
<img width="1050" alt="image" src="https://user-images.githubusercontent.com/3121201/181775087-d9ec2eb3-da28-42be-b543-cfd5bc46a1f6.png">
